src/functional/histogram_compression.zig

```diff
@@ -0,0 +1,519 @@
+    fn minMerge(self: *Self) !void {
+
+        // Pop the smallest merge error (the least costly merge).
+        const min_merge_error: MergeError = try self.merge_queue.remove();
+
+        // Merge the buckets at min_merge_error.index and min_merge_error.index + 1.
+        const index = min_merge_error.index;
+        const bucket_one = self.buckets.items[index];
+        const bucket_two = self.buckets.items[index + 1];
+
+        // Merge the two buckets, updating their range and min/max values.
+        const new_begin = bucket_one.begin;
+        const new_end = bucket_two.end;
+        const new_min = @min(bucket_one.min_val, bucket_two.min_val);
+        const new_max = @max(bucket_one.max_val, bucket_two.max_val);
+
+        // Replace the first bucket with the merged result and remove the second.
+        self.buckets.items[index] = try Bucket.init(new_begin, new_end, new_min, new_max);
+        _ = self.buckets.orderedRemove(index + 1); // Remove the merged bucket.
+
+        if (index < self.buckets.items.len - 1) {
+            const new_merge_error = self.calculateMergeError(index);
+            try self.merge_queue.add(MergeError{ .index = index, .merge_error = new_merge_error });
+
+            if (index > 0)
+                try self.merge(index - 1); // Update error for the previous bucket.
+
+            try self.updateAllIndex(index + 1);
+        }
+    }
+
+    /// Update the merge error for the bucket at 'index'.
+    fn merge(self: *Self, index: usize) !void {
+        const merge_error = self.calculateMergeError(index);
+        const new_merge_error = MergeError{ .index = index, .merge_error = merge_error };
+
+        // Placeholder for the old merge.
+        const old_merge_item: MergeError = MergeError{ .index = index, .merge_error = merge_error };
+
+        // Update the priority queue with the new merge error using the 'update' method.
+        try self.merge_queue.update(old_merge_item, new_merge_error);
+    }
+
+    /// Update the indices of all merge error from `index` until `buckets.items.len`.
+    fn updateAllIndex(self: *Self, index: usize) !void {
+        for (index..self.buckets.items.len) |i| {
+            const merge_error_index: usize = try self.merge_queue.getItemPosition(
+                .{ .index = i, .merge_error = 0 },
+            );
+            const old_merge_error: MergeError = try self.merge_queue.getItemAt(merge_error_index);
+            var new_merge_error: MergeError = .{
+                .index = old_merge_error.index - 1,
+                .merge_error = old_merge_error.merge_error,
+            };
+
+            if (i == index) {
+                new_merge_error.index = self.max_buckets + 10;
+                new_merge_error.merge_error = 1e16;
+            }
+            try self.merge_queue.update(old_merge_error, new_merge_error);
+        }
+    }
+};
+
+/// Compress `uncompressed_values` with the maximum number of buckets defined by the `error_bound`
```

Comment:
No this is weird, we need to modify `tersets.zig` and `capi.zig` to accept different types of parameters like `error_bounds` , `max_buckets`, or even `compression_ratio` directly as some methods accept this parameter directly.

Reply:
If you agree, I can open a new PR for this!

Reply:
I am not sure I fully understand you comment, do you suggest that we change the interface in `tersets.zig` and `capi.zig` to accept different ways for specifying how much error is allowed based on what method is used? I think that is a good idea as it is becoming clear that simply calling everything `error_bound` is not general enough for all methods and is limited to a single value with a single type. But I agree that it is probably too big a change to do in this PR, so I also think it is a good idea to create an issue where we can discuss exactly how to change the interface and then a followup PR to implement it. One idea could maybe be to change `Method` somehow so each compression method can include information and then move `error_bound` into it. This would make it very clear for the user what needs to be set for each compression method and would allow us to specify different parameters for each compression method. For example, maybe `PoorMansCompressionMidrange` could somehow contain its error bound as a single `f32` named something like `absolute_per_value_error_bound`. But of course this should be discussed further before a decision is made due to how big an impact such a change would have.

Reply:
I think you fully understand. I agree this needs proper planning in a meeting to avoid unnecessary iterations. Nevertheless, I will answer the question.
Q1. Do you suggest that we change the interface in `tersets.zig` and `capi.zig` to accept different ways for specifying how much error is allowed based on what method is used? 
A1: Close but a bit different. I don't mean how much error is allowed but instead what kind of error is allowed. For example, all Line-Simplification-based and Histogram-based methods can accept as a parameter that specifies the amount of compression ratio (CR) instead, or completely disregarding the decompression error. In the case of PWCH, it is proved that it is "optimal", meaning that for that amount of CR (using histograms), the amount of error is minimum. Right now, we don't accept that kind of parameter and assume that everything is some kind of error. Thus, we need to modify `tersets.zig` and `capi.zig` to accept that different kind of parameter.